### PR TITLE
Initial support for Terraform v0.12.x

### PR DESCRIPTION
**Pull request by ketzacoatl**
Thanks for @Magicloud for tackling the repo-wide upgrade! The MR has been testing well, but it's likely imperfect. I'd rather release and fix and release updates, than hold this back. 

- [X] Update the changelog
- [X] Make sure that the linting passes on CI.

### CODE REVIEW

**Comment by Magicloud** [comment]
I will make the changelog.

For lint issue, where did you get it? https://gitlab.fpcomplete.com/fpco-mirrors/terraform-aws-foundation/-/jobs/12750 is the 0.12 branch built with new default-build-image. Lint seems passed.

**Comment by Magicloud** [comment]
One thing I am not sure is the hardcoded version of terraform in Makefile. Why 0.12.2 when latest release is 0.12.4? I think this also affects default-build-image. If we want to stick to a non-latest version, I'd suggest to have a KV DB to store those kind of "company preference". So every where, it is the right version.

**Comment by Magicloud** [comment]
The review system is bit of weird, so I post some comments without a link to source.

In _examples/load-asg/main.tf_, around line 105, you removed the TODO generated by TF. That is fine. But seems like you also removed two other comments, started with "Ensure", and "ELBs". Just to confirm it is on purpose.

**Comment by Magicloud** [comment]
And there are multiple syntax error, for example, _examples/kube-stack-public/kube.tf_, line 24.

It is a lint for one element list. But the closing bracket is gone.

**Comment by ketzacoatl** [comment]
Is the code review physically not working? It would be helpful if you could put the comments at the relevant code in the future. Thank you for the review, and you are correct, removing the TODO should not have removed the next line.

**Comment by Magicloud** [comment]
Sorry, get it working now. Posted my comments.

And for syntax error, I think I made mistake. I said multiple, but reviewing today gets only one.

File: modules/single-port-sg/main.tf [hunk 1]
```diff
@@ -46,24 +46,24 @@ locals {
...
-  count             = "${local.tcp}"
+  count             = local.tcp
   type              = "ingress"
   description       = "${var.description} (tcp)"
-  from_port         = "${var.port}"
-  to_port           = "${var.port}"
+  from_port         = var.port
+  to_port           = var.port
   protocol          = "tcp"
   cidr_blocks       = ["${var.cidr_blocks}"]
```
Comment (qrilka) [inline modules/single-port-sg/main.tf]: this one should change as well

Comment (ketzacoatl) [reply modules/single-port-sg/main.tf]: Yes, definitely. Good catch, I have updated that.

**Comment by ketzacoatl** [comment]
@qrilka / @Magicloud ok, I have updated the branch again. Please review one last time, give approval if that is ready, and we can merge.

File: examples/gitlab-ha/main.tf [hunk 2]
```diff
@@ -21,13 +21,14 @@
...
+  region = var.region
 }
 
-data "aws_availability_zones" "available" {}
+data "aws_availability_zones" "available" {
+}
 
 data "aws_route53_zone" "zone" {
-  name = "${var.dns_zone_name}."
+  name         = "${var.dns_zone_name}."
```
Comment (psibi) [inline examples/gitlab-ha/main.tf]: The trailing dot seems to be accidental here.

File: modules/bind-server/main.tf [hunk 3]
```diff
@@ -27,73 +27,73 @@ resource "aws_instance" "bind" {
...
 
-  tags {
-    Name = "${length(var.names) == 0 ? format("%s-%02d", var.name, count.index + 1) : var.names[count.index]}"
+  tags = {
+    Name = length(var.names) == 0 ? format("%s-%02d", var.name, count.index + 1) : var.names[count.index]
   }
 
   lifecycle {
-    ignore_changes = ["key_name"]
+    ignore_changes = [key_name]
```
Comment (psibi) [inline modules/bind-server/main.tf]: Is this right or should it be `var.key_name` ?

File: examples/gitlab-ha/main.tf (hunk 2 above)
Comment (ketzacoatl) [reply examples/gitlab-ha/main.tf]: I believe it was added assuming that the operator would provide `foo.bar.tld`, but we'd need to query with `foo.bar.tld.` - this is a DNS thing

File: modules/bind-server/main.tf (hunk 3 above)
Comment (ketzacoatl) [reply modules/bind-server/main.tf]: I think it should be `"key_name"`, good catch.

**Pull request merged**